### ISSUE TITLE
Add a package manager

### ISSUE THREAD

**Issue by Anonymous3-a** [nushell/nushell#3734]

**Comment by skelly37** [comment]
I'm starting to work on it and I invite anyone interested in collaboration to do so

https://github.com/skelly37/nupac

**Comment by adamkpickering** [comment]
I think this is a great idea. However, I have one concern: dependency management.

When I was using Ansible, I often had a hard time due to dependencies. Ansible is kind of weird in that it can get dependencies from three different package managers:

- Ansible Galaxy, which provides pre-packaged Ansible "modules" (roles and collections)
- pip, which provides Python packages
- whatever the system package manager is for the distro you're using (`apt`, `yum` etc)

Basically you could have an Ansible role that required certain Python packages, and those Python packages could require certain libraries to be installed on your system by your system package manager. And none of the package managers are aware of the others except at runtime. As you can imagine, headaches are the result.

So a word from the (I like to think) wise: if nushell gets a package manager, **there need to be rules around what can go into a package**. Specifically, I'd propose that **a package should have no external dependencies**. Ideally it should be a statically linked binary. This means that users will be able to download a package and expect it to just work, without having to worry about what OS they're on, what they have installed and so on. This would greatly simplify users' lives.

**Comment by skelly37** [comment]
> I think this is a great idea. However, I have one concern: dependency management.
> 
> When I was using Ansible, I often had a hard time due to dependencies. Ansible is kind of weird in that it can get dependencies from three different package managers:
> 
>     * Ansible Galaxy, which provides pre-packaged Ansible "modules" (roles and collections)
> 
>     * pip, which provides Python packages
> 
>     * whatever the system package manager is for the distro you're using (`apt`, `yum` etc)
> 
> 
> Basically you could have an Ansible role that required certain Python packages, and those Python packages could require certain libraries to be installed on your system by your system package manager. And none of the packages say which packages from the underlying one they need except for in documentation. As you can imagine, runtime errors, requiring docker containers to run scripts, willy-nilliy installing/updating of packages, and headaches are the result.
> 
> So a word from the (I like to think) wise: if nushell gets a package manager, **there need to be rules around what can go into a package**. Specifically, I'd propose that **a package should have no external dependencies**. Ideally it should be a statically linked binary. This means that users will be able to download a package and expect it to just work, without having to worry about what OS they're on, what they have installed and so on. This would greatly simplify users' lives.

I agree on that it should *just work*. Handling nu_scripts is of no concern and I think we should start out with it (that's my plan right after the release). 

Releasing alpha version working on nu_scripts would be a great idea, because such project would be rather easy to write and we could get feedback on the user interface. When the interface is polished, we can work on the more advanced tasks, worrying just about the logic, repo rules and plugin manifests.

Also, my ideas are stated in the mentioned nupac repo, so feel free to comment or discuss them.
